src/Features/LanguageServer/Protocol/Handler/ExportLspRequestHandlerProviderAttribute.cs

```diff
@@ -15,25 +16,38 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     [AttributeUsage(AttributeTargets.Class), MetadataAttribute]
     internal class ExportLspRequestHandlerProviderAttribute : ExportAttribute
     {
+        public Type[] HandlerTypes { get; }
+
         /// <summary>
-        /// The document languages that this handler supports.
+        /// Exports an <see cref="AbstractRequestHandlerProvider"/> and specifies the contract and
+        /// <see cref="IRequestHandler"/> types this provider is associated with.
         /// </summary>
-        public string[] LanguageNames { get; }
-
-        public ExportLspRequestHandlerProviderAttribute(params string[] languageNames) : base(typeof(AbstractRequestHandlerProvider))
+        /// <param name="contractName">
+        /// The contract name this provider is exported.  Used by <see cref="AbstractRequestDispatcherFactory"/>
+        /// when importing handlers to ensure that it only imports handlers that match this contract.
+        /// This is important to ensure that we only load relevant providers (e.g. don't load Xaml providers when creating the c# server),
+        /// otherwise we will get dll load RPS regressions for the <see cref="HandlerTypes"/>
+        /// </param>
+        /// <param name="firstHandlerType">
+        /// The concrete type of the <see cref="IRequestHandler"/> provided in <see cref="AbstractRequestHandlerProvider.CreateRequestHandlers(WellKnownLspServerKinds)"/>
+        /// </param>
+        /// <param name="additionalHandlerTypes">
+        /// Additional <see cref="IRequestHandler"/> if <see cref="AbstractRequestHandlerProvider.CreateRequestHandlers(WellKnownLspServerKinds)"/>
+        /// provides more than one handler at once.
+        /// </param>
+        public ExportLspRequestHandlerProviderAttribute(string contractName, Type firstHandlerType, params Type[] additionalHandlerTypes) : base(contractName, typeof(AbstractRequestHandlerProvider))
```

Comment:
the firstHandler and additionalHandlerTypes is so we get a compile error if you don't pass in at least one handler type